### PR TITLE
Testing

### PR DESCRIPTION
**Pull request by OmriBarZik**

### CODE REVIEW

File: tests/phpunit/manager.php [hunk 1]
```diff
@@ -0,0 +1,30 @@
+<?php
+
+namespace Elementor\Testing;
+
+class Manager {
+	/**
+	 * @var self
+	 */
+	public static $instance = null;
```
Comment (KingYes) [inline tests/phpunit/manager.php]: private

File: tests/bootstrap.php [hunk 2]
```diff
@@ -23,30 +24,43 @@
...
 	return $html;
 }
 
 require $_tests_dir . '/includes/bootstrap.php';
+require 'phpunit/local-factory.php';
+require 'phpunit/base-class.php';
+require 'phpunit/manager.php';
+\Elementor\Testing\Manager::instance();
+
+global $argv;
```
Comment (KingYes) [inline tests/bootstrap.php]: Try use with Bootstrap class

File: tests/phpunit/elementor/includes/base/test-controls-stack.php [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+
+	public function test_should_get_current_section() {
+		self::assertEquals( $this->manager->get_current_section(), '' );
+	}
+
+	public function test_should_get_active_controls() {
+		self::assertEquals( $this->manager->get_active_controls(), [] );
+	}
+
+	public function test_Should_add_control() {
```
Comment (KingYes) [inline tests/phpunit/elementor/includes/base/test-controls-stack.php]: spell

File: tests/phpunit/elementor/includes/template-library/test-manager-general.php [hunk 4]
```diff
@@ -0,0 +1,213 @@
...
+	public static function setUpBeforeClass() {
+		self::$manager = \Elementor\Plugin::instance()->templates_manager;
+	}
+
+	public function test_should_return_import_images_instance() {
+		$this->assertEquals( self::$manager->get_import_images_instance(), new \Elementor\TemplateLibrary\Classes\Import_Images() );
+	}
+
+	public function test_should_return_wp_error_source_class_name_not_exists_from_register_source() {
+		$this->assertWPError( self::$manager->register_source( 'pop' ), 'source_class_name_not_exists' );
```
Comment (KingYes) [inline tests/phpunit/elementor/includes/template-library/test-manager-general.php]: change `pop`

File: tests/phpunit/elementor/includes/template-library/test-manager-general.php [hunk 5]
```diff
@@ -0,0 +1,213 @@
...
+	}
+
+	public function test_should_return_wp_error_wrong_instance_source_from_register_source() {
+		$this->assertWPError( self::$manager->register_source( 'Elementor\Core\Ajax_Manager' ), 'wrong_instance_source' );
+	}
+
+
+	public function test_should_return_false_from_unregister_source() {
+		$this->assertFalse( self::$manager->unregister_source( 0 ) );
```
Comment (KingYes) [inline tests/phpunit/elementor/includes/template-library/test-manager-general.php]: change `0`.

File: tests/phpunit/elementor/includes/template-library/test-manager-general.php [hunk 6]
```diff
@@ -0,0 +1,213 @@
...
+	}
+
+	public function test_should_fail_to_return_source() {
+		$this->assertFalse( self::$manager->get_source( 'pop' ) );
+	}
+
+	public function mockGetTemplate() {
```
Comment (KingYes) [inline tests/phpunit/elementor/includes/template-library/test-manager-general.php]: rename & move & protected

File: tests/phpunit/elementor/includes/template-library/test-manager-general.php [hunk 7]
```diff
@@ -0,0 +1,213 @@
...
+		foreach ( $source_array as $source ) {
+			$templates_array = array_merge( $templates_array, $source->get_items() );
+		}
+
+		return $templates_array;
+	}
+
+	public function test_should_return_templates() {
+		//run & check.
+		self::$manager->register_source( 'Elementor\TemplateLibrary\Source_Remote' );
```
Comment (KingYes) [inline tests/phpunit/elementor/includes/template-library/test-manager-general.php]: why register?

File: tests/phpunit/elementor/includes/template-library/test-manager-general.php [hunk 8]
```diff
@@ -0,0 +1,213 @@
...
+		self::$manager->register_source( 'Elementor\TemplateLibrary\Source_Local' );
+		$this->assertEquals( self::$manager->get_templates(), $this->mockGetTemplate() );
+	}
+
+	public function test_should_return_library_data() {
+		self::$manager->register_source( 'Elementor\TemplateLibrary\Source_Remote' );
```
Comment (KingYes) [inline tests/phpunit/elementor/includes/template-library/test-manager-general.php]: why register?

File: tests/phpunit/elementor/includes/template-library/test-manager-general.php [hunk 9]
```diff
@@ -0,0 +1,213 @@
...
+		self::$manager->register_source( 'Elementor\TemplateLibrary\Source_Local' );
+
+		$ret = self::$manager->get_library_data( [] );
+
+		$this->assertNotEmpty( $ret['templates'] );
+	}
+
+	public function test_should_return_library_data_send_this_parameters() {
```
Comment (KingYes) [inline tests/phpunit/elementor/includes/template-library/test-manager-general.php]: remove

File: tests/phpunit/elementor/includes/template-library/test-manager-general.php [hunk 10]
```diff
@@ -0,0 +1,213 @@
...
+		self::$manager->register_source( 'Elementor\TemplateLibrary\Source_Remote' );
+		self::$manager->register_source( 'Elementor\TemplateLibrary\Source_Local' );
+
+		$ret = self::$manager->get_library_data( [ 'sync' => true ] );
+
+		$this->assertNotEmpty( $ret['templates'] );
+	}
+
+	public function test_should_return_wp_error_arguments_not_specified_from_save_template() {
+		$this->assertWPError(
```
Comment (KingYes) [inline tests/phpunit/elementor/includes/template-library/test-manager-general.php]: `$fake_post_id = '123';`

File: tests/phpunit/elementor/includes/template-library/test-manager-general.php [hunk 11]
```diff
@@ -0,0 +1,213 @@
...
+					'id' => 1,
+				],
+			],
+		];
+		$this->assertWPError( self::$manager->update_templates( $templates ) );
+	}
+
+	public function test_should_return_true_from_update_templates() {
+		wp_set_current_user( $this->factory()->create_and_get_administrator_user()->ID );
+		$templates = [
```
Comment (KingYes) [inline tests/phpunit/elementor/includes/template-library/test-manager-general.php]: line

File: tests/phpunit/elementor/includes/template-library/test-manager-general.php [hunk 12]
```diff
@@ -0,0 +1,213 @@
...
+			'templates' => [
```
Comment (KingYes) [inline tests/phpunit/elementor/includes/template-library/test-manager-general.php]: replace false to WP_Error

File: tests/phpunit/elementor/includes/template-library/test-manager-general.php [hunk 13]
```diff
@@ -0,0 +1,213 @@
...
+			self::$manager->export_template(
+				[
+					'source' => 'banana',
+					'template_id' => '777',
+				]
+			), 'template_error'
+		);
+	}
+
+	/*    public function () {
```
Comment (KingYes) [inline tests/phpunit/elementor/includes/template-library/test-manager-general.php]: cleanup

File: tests/phpunit/elementor/includes/template-library/test-manager-local.php [hunk 14]
```diff
@@ -0,0 +1,212 @@
+<?php /** @noinspection PhpParamsInspection */
```
Comment (KingYes) [inline tests/phpunit/elementor/includes/template-library/test-manager-local.php]: ?

File: tests/phpunit/elementor/includes/template-library/test-manager-local.php [hunk 15]
```diff
@@ -0,0 +1,212 @@
...
+
+namespace Elementor\Testing\Includes\TemplateLibrary;
+
+use \Elementor\TemplateLibrary\Manager;
+use Elementor\Testing\Elementor_Test_Base;
+
+require_once 'test-manager-general.php';
```
Comment (KingYes) [inline tests/phpunit/elementor/includes/template-library/test-manager-local.php]: remove

File: tests/phpunit/elementor/includes/template-library/test-manager-local.php [hunk 16]
```diff
@@ -0,0 +1,212 @@
...
+		self::$manager = \Elementor\Plugin::instance()->templates_manager;
+	}
+
+	public function setUp() {
+		parent::setUp();
+		wp_set_current_user( $this->factory()->get_administrator_user()->ID );
+	}
+
+	public function test_should_return_true_from_register_source() {
+		$this->assertTrue( self::$manager->register_source( 'Elementor\TemplateLibrary\Source_Local' ) );
```
Comment (KingYes) [inline tests/phpunit/elementor/includes/template-library/test-manager-local.php]: add condition ensuring is exist.

File: tests/phpunit/elementor/includes/template-library/test-manager-local.php [hunk 17]
```diff
@@ -0,0 +1,212 @@
...
+	}
+
+	public function test_should_return_true_from_unregister_source() {
```
Comment (KingYes) [inline tests/phpunit/elementor/includes/template-library/test-manager-local.php]: swap

File: tests/phpunit/elementor/includes/template-library/test-manager-local.php [hunk 18]
```diff
@@ -0,0 +1,212 @@
...
+		$this->assertTrue( self::$manager->unregister_source( 'local' ) );
+	}
+
+	public function test_should_return_registered_sources() {
+		self::$manager->register_source( 'Elementor\TemplateLibrary\Source_Local' );
```
Comment (KingYes) [inline tests/phpunit/elementor/includes/template-library/test-manager-local.php]: remove this line

File: tests/phpunit/elementor/includes/template-library/test-manager-local.php [hunk 19]
```diff
@@ -0,0 +1,212 @@
...
+			), 'save_error'
+		);
+	}
+
+	/**
+	 *
+	 */
+	public function test_should_return_template_data_from_update_template() {
+		wp_set_current_user( $this->factory()->create_and_get_administrator_user()->ID );
+		$post_id = $this->factory()->create_and_get_default_Post()->ID;
```
Comment (KingYes) [inline tests/phpunit/elementor/includes/template-library/test-manager-local.php]: upper?

File: tests/phpunit/elementor/includes/template-library/test-manager-remot.php [hunk 20]
```diff
@@ -0,0 +1,123 @@
...
+			'tags',
+			'isPro',
+			'hasPageSettings',
+			'url',
+			'favorite',
+		];
+		$this->assertArrayHaveKeys( $remote_remote, self::$manager->update_template( $template_data ) );
+	}
+
+	public function test_should_return_data_from_get_template_data() {
```
Comment (KingYes) [inline tests/phpunit/elementor/includes/template-library/test-manager-remot.php]: ?

File: tests/phpunit/elementor/includes/template-library/test-manager-remot.php [hunk 21]
```diff
@@ -0,0 +1,123 @@
...
+			[
+				'source' => 'local',
+				'template_id' => '777',
+			]
+		);
+
+		$this->assertEquals( $ret, [ 'content' => [] ] );
+	}
+
+	public function test_should_delete_template() {
```
Comment (KingYes) [inline tests/phpunit/elementor/includes/template-library/test-manager-remot.php]: ?

**Review by KingYes — CHANGES_REQUESTED** [review]

File: tests/phpunit/manager.php (hunk 1 above)
Comment (OmriBar-Zik) [reply tests/phpunit/manager.php]: done

File: tests/phpunit/elementor/includes/base/test-controls-stack.php (hunk 3 above)
Comment (OmriBar-Zik) [reply tests/phpunit/elementor/includes/base/test-controls-stack.php]: done

File: tests/phpunit/elementor/includes/template-library/test-manager-general.php [hunk 22]
```diff
@@ -10,11 +10,23 @@ class Elementor_Test_Manager_General extends Elementor_Test_Base {
 	 * @var Manager
 	 */
 	protected static $manager;
+	private $fake_post_id = '123';
 
 	public static function setUpBeforeClass() {
 		self::$manager = \Elementor\Plugin::instance()->templates_manager;
 	}
 
+	public function mockGetTemplate() {
```
Comment (KingYes) [inline tests/phpunit/elementor/includes/template-library/test-manager-general.php]: rename

File: includes/template-library/sources/remote.php [hunk 23]
```diff
@@ -1,4 +1,5 @@
 <?php
+
```
Comment (KingYes) [inline includes/template-library/sources/remote.php]: remove line

File: includes/template-library/sources/remote.php [hunk 24]
```diff
@@ -55,7 +56,7 @@ public function get_title() {
 	 * @since 1.0.0
 	 * @access public
 	 */
-	public function register_data() {}
+	public function register_data() { }
```
Comment (KingYes) [inline includes/template-library/sources/remote.php]: space?

File: includes/template-library/sources/remote.php [hunk 25]
```diff
@@ -117,10 +118,10 @@ public function get_item( $template_id ) {
 	 *
 	 * @param array $template_data Remote template data.
 	 *
-	 * @return bool Return false.
+	 * @return \WP_Error
 	 */
 	public function save_item( $template_data ) {
-		return false;
+		return new \WP_Error( 'invalid_request', 'cannot save template from remote source' );
```
Comment (KingYes) [inline includes/template-library/sources/remote.php]: from => to

File: tests/phpunit/elementor/includes/template-library/test-manager-remot.php [hunk 26]
```diff
@@ -111,13 +81,13 @@ public function test_should_return_data_from_get_template_data() {
...
-					'source' => 'remote',
-					'template_id' => '777',
-				]
-			)
-		);
+
+		$this->assertWPError( self::$manager->delete_template(
+			[
+				'source' => 'remote',
+				'template_id' => '777',
```
Comment (KingYes) [inline tests/phpunit/elementor/includes/template-library/test-manager-remot.php]: id?

File: tests/phpunit/elementor/includes/template-library/test-manager-remote.php [hunk 27]
```diff
@@ -45,7 +45,7 @@ public function test_should_return_source() {
 		$this->assertEquals( self::$manager->get_source( 'remote' ), new \Elementor\TemplateLibrary\Source_Remote() );
 	}
 
-	public function test_should_return_template_data_from_save_template() {
+	public function test_should_return_WP_Error_from_save_template() {
```
Comment (KingYes) [inline tests/phpunit/elementor/includes/template-library/test-manager-remote.php]: lower

**Review by KingYes — CHANGES_REQUESTED** [review]

File: tests/phpunit/elementor/includes/template-library/test-manager-general.php (hunk 4 above)
Comment (KingYes) [reply tests/phpunit/elementor/includes/template-library/test-manager-general.php]: ?

File: tests/phpunit/elementor/includes/template-library/test-manager-local.php (hunk 15 above)
Comment (KingYes) [reply tests/phpunit/elementor/includes/template-library/test-manager-local.php]: ?

File: tests/phpunit/elementor/includes/template-library/test-manager-remote.php [hunk 28]
```diff
@@ -0,0 +1,93 @@
...
+		$this->assertEquals( self::$manager->get_registered_sources()['remote'], new \Elementor\TemplateLibrary\Source_Remote() );
+	}
+
+	public function test_should_return_source() {
+		$this->assertEquals( self::$manager->get_source( 'remote' ), new \Elementor\TemplateLibrary\Source_Remote() );
+	}
+
+	public function test_should_return_wp_error_from_save_template() {
+		$template_data = [
+			'post_id' => '777',
```
Comment (KingYes) [inline tests/phpunit/elementor/includes/template-library/test-manager-remote.php]: ?

File: tests/phpunit/elementor/includes/template-library/test-manager-remote.php [hunk 29]
```diff
@@ -0,0 +1,93 @@
...
+	public static function setUpBeforeClass() {
+		self::$manager = \Elementor\Plugin::instance()->templates_manager;
+	}
+
+	public function test_should_mark_template_as_favorite() {
+		$this->assertFalse(
+			self::$manager->mark_template_as_favorite(
+				[
+					'source' => 'remote',
+					'template_id' => '777',
```
Comment (KingYes) [inline tests/phpunit/elementor/includes/template-library/test-manager-remote.php]: ?

**Review by KingYes — CHANGES_REQUESTED** [review]

**Pull request closed without merging**